File: Java/other/DoubleFactorial.java
import java.util.*;

public class DoubleFactorial {

    // Recursive approach
    private static int recursiveDoubleFactorial(int num) {
        // checking for 1 and 0
        if (num == 0 || num == 1) {
            return 1;
        }

        // using recursion
        return num * recursiveDoubleFactorial(num - 2);
    }

    // Iterative approach
    private static int iterDoubleFactorial(int num) {
        int result = 1;
        // using iterations
        for (int i = num; i >= 0; i = i-2) {

            if (i == 0 || i == 1) {
                return result;
            } else {
                result *= i;
            }
        }

        return result;
    }
    
    public static void main(String[] args) {
        
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter the number:");
        int num = sc.nextInt();
        System.out.println("Which approach would you like to use?");
        System.out.println("1. Recursive");
        System.out.println("2. Iterative");
        int ch = sc.nextInt();
        sc.close();

        int result;
        switch(ch) {
            case 1:
                result = recursiveDoubleFactorial(num);
                System.out.println("Double factorial:" + result);
                break;

            case 2:
                result = iterDoubleFactorial(num);
                System.out.println("Double factorial:" + result);
                break;

            default:
                System.out.println("Invalid choice.");
        }
    }
}
/**
 * Sample input/output
 * Enter the number:
 * 5
 * Which approach would you like to use?
 * 1. Recursive
 * 2. Iterative
 * 1
 * Double factorial:15
 * 
 * Enter the number:
 * 6
 * Which approach would you like to use?
 * 1. Recursive
 * 2. Iterative
 * 2
 * Double factorial:48
 * 
 * Time complexity for both the approaches is O(n)
 */

File: C/cp/doublefactorial.c
#include <stdio.h>

int iterative_doublefactorial(int n) {
  int res = 1;
  for (int i = n; i > 0; i -= 2) {
    if (n == 0 || n == 1)
      return res;
    else
      res = res * i;
  }

}

int recursive_doublefactorial(int n) {
  if (n == 0 || n == 1)
    return 1;
  else
    return n * recursive_doublefactorial(n - 2);
}

int main() {
  int n;
  printf("Enter the number :");
  scanf("%d", & n);
  int k;
  do {
    printf("Enter the choice below :\n1.Iterative solution \n2.Recursive Solution \n3.Exit \n");
    scanf("%d", & k);
    switch (k) {
    case 1:
      printf("The Double Factorial is : %d\n", iterative_doublefactorial(n));
      break;
    case 2:
      printf("The Double Factorial is : %d\n", recursive_doublefactorial(n));
      break;
    }
  } while (k != 3);

  return 0;
}

/*
Enter the number :7
Enter the choice below :
1.Iterative solution 
2.Recursive Solution 
3.Exit 
1
The Double Factorial is : 105
Enter the choice below :
1.Iterative solution 
2.Recursive Solution 
3.Exit 
2
The Double Factorial is : 105
Enter the choice below :
1.Iterative solution 
2.Recursive Solution 
3.Exit 
3
*/
